keps/sig-api-machinery/4355-coordinated-leader-election/README.md

```diff
@@ -0,0 +1,966 @@
+    C --> |Renew Leader Lease| C
+    C -->|End of Term / Leader Lease Expired| D[Shutdown]
+    D[Shutdown] -.-> |Restart| A
+```
+
+### Enabling on a component
+
+Components with a `--leader-elect-resource-lock` flag (kube-controller-manager,
+ kube-scheduler) will accept `coordinatedleases` as a resource lock type.
+
+### Comparison of leader election
+
+|                 | Lease Based Leader Election      | Coordinated Leader Election                                                    |
+| --------------- | -------------------------------- | ------------------------------------------------------------------------------ |
+| Lock Type       | Lease                            | Lease                                                                          |
+| Claimed by      | Component instance               | Election Coordinator. (Lease is claimed for to the elected component instance) |
+| Renewed by      | Component instance               | Component instance                                                             |
+| Leader Criteria | First component to claim lease   | Best leader from available candidates at time of election                      |
+| Preemptable     | No                               | Yes, Collaboratively. (Coordinator marks lease as "end of term". Component instance voluntarily stops renewing) |
+
+### User Stories (Optional)
+
+#### Story 1
+
+A cluster administrator upgrades a cluster's control plane node-by-node,
+expecting version skew to be respected.
+
+- When the first and second nodes are upgraded, any components that were leaders
+  will typically loose the lease during the node downtime
+  - If one happens to retain it's lease, it will be preempted by the coordinated
+    election controller after it updates it's identity lease with new version
+    information
+- When the third node is upgraded, all components will be at the new version and one
+  will be elected
+
+#### Story 2
+
+A cluster administrator rolls back a cluster's control plane node-by-node,
+expecting version skew to be respected.
+
+- When the first node is rolled back, any components that were leaders
+  will typically loose the lease during the node downtime
+- Once one of the components updates its identity lease with new version
+  information, the coordinated election controller will preempt the current
+  leader so that this lower version component becomes leader.
+- When the remaining two nodes can rollback, the first node will typically
+  remain leader, but if a new election occurs, the available older version
+  components will be elected.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+#### Risk: This breaks leader election in some super subtle way
```

Comment:
Can you describe those "subtle ways"?

Reply:
Sure, I was thinking about the "We have to "start over" and build confidence in a new leader election algorithm" problem here. I'll write it out.